skimage/restoration/_nl_means_denoising.pyx

```diff
@@ -0,0 +1,712 @@
+    for pln in range(offset, n_pln + offset):
+        pln_start = pln - offset
+        pln_end = pln + offset + 1
+        # Iterate over rows, taking padding into account
+        for row in range(offset, n_row + offset):
+            row_start = row - offset
+            row_end = row + offset + 1
+            # Iterate over columns, taking padding into account
+            for col in range(offset, n_col + offset):
+                col_start = col - offset
+                col_end = col + offset + 1
+                new_value = 0
+                weight_sum = 0
+
+                # Iterate over local 3d patch for each pixel
+                # First planes
+                for i in range(max(-d, offset - pln),
+                               min(d + 1, n_pln + offset - pln)):
+                    pln_start_i = pln_start + i
+                    pln_end_i = pln_end + i
+                    # Rows
+                    for j in range(max(-d, offset - row),
+                                   min(d + 1, n_row + offset - row)):
+                        row_start_j = row_start + j
+                        row_end_j = row_end + j
+                        # Columns
+                        for k in range(max(-d, offset - col),
+                                       min(d + 1, n_col + offset - col)):
+                            col_start_k = col_start + k
+                            col_end_k = col_end + k
+                            weight = patch_distance_3d(
+                                    padded[pln_start:pln_end,
+                                           row_start:row_end,
+                                           col_start:col_end],
+                                    padded[pln_start_i:pln_end_i,
+                                           row_start_j:row_end_j,
+                                           col_start_k:col_end_k],
+                                    w, s)
+                            # Collect results in weight sum
+                            weight_sum += weight
+                            new_value += weight * padded[pln + i,
+                                                         row + j, col + k]
+
+                # Normalize the result
+                result[pln, row, col] = new_value / weight_sum
+
+    # Return cropped result, undoing padding
+    return result[offset:-offset, offset:-offset, offset:-offset]
+
+#-------------- Accelerated algorithm of Froment 2015 ------------------
+
+
+@cython.cdivision(True)
+@cython.boundscheck(False)
+cdef inline float _integral_to_distance_2d(IMGDTYPE [:, ::] integral,
+                        int row, int col, int offset, float h2s2):
+    """
+    References
+    ----------
+    Jacques Froment. Parameter-Free Fast Pixelwise Non-Local Means
+    Denoising. Image Processing On Line, 2014, vol. 4, p. 300-326.
+
+    Used in _fast_nl_means_denoising_2d
+    """
+    cdef float distance
+    distance =  integral[row + offset, col + offset] + \
```

Comment:
Would precalculating these offsets save half the addition/subtraction operations? Or is Cython smart enough to do this without being told? Might be worthwhile, as it's the innermost loop.

``` python
row_pos = row + offset
row_neg = row - offset
col_pos = col + offset
col_neg = col - offset
```

Doing this for 3d in the next function actually cuts out 75% of the simple math ops.